### PR TITLE
Modularity improvements, part II

### LINKED ISSUE #540: Extend the module API with error reporting
So it has happened to me just today that a module initialization failure torn down the whole daemon. Calling `udisks_error()` from modules is really not a good idea unless there's real reason for that. Citing the docs:

> \* udisks_error:
> \* \@args...: printf()-style format string and arguments
> \*
> \* Logging macro for %UDISKS_LOG_LEVEL_ERROR.
> \* Warning: THIS ALSO TERMINATES THE PROCESS with the abort() call!

In particular this comes from the `UDisksModuleInitFunc` implementation on a module side after `bd_reinit()` fails. That explains the
> /* XXX: can do nothing more here even though we know the module will be unusable! */

comments.

For an uknown reason none of the UDisksModule* funcs are taking GError** argument for error reporting and with `UDisksModuleInitFunc` the return value is even used for returning module state pointer. Since I wrote the modular API I take full responsibility for that mistake. :-)

----

While there are no defined API guarantees fortunately a warning is noted down in the docs:

> there's no support for out-of-the-tree modules and care must be taken when changing UDisks internals.

So my proposal is to revise and extend at least the `UDisksModuleInitFunc` API and adapting the in-tree modules accordingly.

Does this sound okay with you?

### PR DESCRIPTION
**Pull request by tbzatek**
Introducing new `UDisksModule` class that should act as a module singleton and a handle, carrying module state and internal structures. This simplifies the module API entry points, most of the functionality was transformed into the `UDisksModule` class methods.

Contains a lot of codestyle fixes to get closer to the unified UDisks conventions.

This should've been done at the beginning, no idea why I had such fear of `GObject` subclassing.

Fixes #540

### CODE REVIEW

**Comment by tbzatek** [comment]
Jenkins, test this please.

**Comment by tbzatek** [comment]
TODO: include module names in test skip messages (LSM)

File: src/udiskslinuxmanager.c [hunk 1]
```diff
@@ -882,11 +883,31 @@ load_modules_in_idle_cb (gpointer user_data)
...
+  else
+    {
+      GError *error = NULL;
 
-  udisks_manager_complete_enable_modules (data->object, data->invocation);
+      /* Load single requested module */
+      if (! udisks_module_manager_load_single_module (module_manager, data->module_name, &error))
+        {
+          g_prefix_error (&error, "Error initializing module '%s': ", data->module_name);
+          g_critical ("%s", error->message);
```
[the remark this answers is not in the document]
Comment (tbzatek) [reply src/udiskslinuxmanager.c:900]: Changed to `g_warning()`. It's not that critical indeed.

File: src/udiskslinuxmanager.c [hunk 2]
```diff
@@ -921,6 +942,37 @@ handle_enable_modules (UDisksManager         *object,
...
+{
+  UDisksLinuxManager *manager = UDISKS_LINUX_MANAGER (object);
+  EnableModulesData *data;
+
+  if (! arg_enable)
+    {
+      /* TODO: implement proper module unloading */
+      g_dbus_method_invocation_return_error_literal (invocation,
+                                                     G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT,
+                                                     "Invalid value \"FALSE\"");
```
[the remark this answers is not in the document]
Comment (tbzatek) [reply src/udiskslinuxmanager.c:959]: Fixed.

File: src/tests/dbus-tests/test_bcache.py [hunk 3]
```diff
@@ -65,6 +65,13 @@ def _get_size(self, bcache_name):
     def _obj_path_from_path(self, device_path):
         return self.path_prefix + '/block_devices/' + os.path.basename(device_path)
 
+    def test__manager_interface(self):
```
[the remark this answers is not in the document]
Comment (tbzatek) [reply src/tests/dbus-tests/test_bcache.py:68]: This was on purpose, to let the test run before others. Pure cosmetic thing.

File: modules/vdo/udiskslinuxblockvdo.c [hunk 4]
```diff
@@ -229,7 +232,8 @@ do_refresh (UDisksBlockVDO *block_vdo,
   udisks_block_vdo_set_name (block_vdo, bd_info->name);
   udisks_block_vdo_set_physical_size (block_vdo, bd_info->physical_size);
   udisks_block_vdo_set_write_policy (block_vdo,
-                                     bd_vdo_get_write_policy_str (bd_info->write_policy, NULL));
+                                     bd_vdo_get_write_policy_str (bd_info->write_policy, &local_error));
+  g_clear_error (&local_error);
```
[the remark this answers is not in the document]
Comment (tbzatek) [reply modules/vdo/udiskslinuxblockvdo.c]: Fixed, reworked a little bit.

File: src/tests/dbus-tests/test_bcache.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vojtechtrefny) [reply src/tests/dbus-tests/test_bcache.py:68]: Ok.

**Comment by tbzatek** [comment]
Tests seem to be stable with 24h runs, apart from segfaulting `lvm` and kernel lockups on Fedora 30 and Fedora 32. Merging.

**Pull request merged**